providers/websockets/huobi/ws_data_handler.go

```diff
@@ -0,0 +1,153 @@
+package huobi
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/klauspost/compress/gzip"
+
+	"go.uber.org/zap"
+
+	"github.com/skip-mev/slinky/oracle/config"
+	"github.com/skip-mev/slinky/providers/base/websocket/handlers"
+	providertypes "github.com/skip-mev/slinky/providers/types"
+	oracletypes "github.com/skip-mev/slinky/x/oracle/types"
+)
+
+var _ handlers.WebSocketDataHandler[oracletypes.CurrencyPair, *big.Int] = (*WebsocketDataHandler)(nil)
+
+// WebsocketDataHandler implements the WebSocketDataHandler interface. This is used to
+// handle messages received from the Huobi websocket API.
+type WebsocketDataHandler struct {
+	logger *zap.Logger
+
+	// config is the config for the Huobi web socket API.
+	cfg config.ProviderConfig
+}
+
+// NewWebSocketDataHandler returns a new WebSocketDataHandler implementation for Huobi
+// from a given provider configuration.
+func NewWebSocketDataHandler(
+	logger *zap.Logger,
+	cfg config.ProviderConfig,
+) (handlers.WebSocketDataHandler[oracletypes.CurrencyPair, *big.Int], error) {
+	if err := cfg.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("invalid provider config %s", err)
+	}
+
+	if !cfg.WebSocket.Enabled {
+		return nil, fmt.Errorf("web socket is not enabled for provider %s", cfg.Name)
+	}
+
+	if cfg.Name != Name {
+		return nil, fmt.Errorf("invalid provider name %s", cfg.Name)
+	}
+
+	return &WebsocketDataHandler{
+		cfg:    cfg,
+		logger: logger.With(zap.String("web_socket_data_handler", Name)),
+	}, nil
+}
+
+// HandleMessage is used to handle a message received from the data provider. The Huobi
+// provider sends two types of messages:
+//
+//  1. Subscribe response message. The subscribe response message is used to determine if
+//     the subscription was successful.
+//  2. Ticker response message. This is sent when a ticker update is received from the
+//     Huobi web socket API.
+//  3. Heartbeat ping message.
+func (h *WebsocketDataHandler) HandleMessage(
+	message []byte,
+) (providertypes.GetResponse[oracletypes.CurrencyPair, *big.Int], []handlers.WebsocketEncodedMessage, error) {
+	var (
+		resp                 providertypes.GetResponse[oracletypes.CurrencyPair, *big.Int]
+		pingMessage          PingMessage
+		subscriptionResponse SubscriptionResponse
+		tickerStream         TickerStream
+	)
+
+	reader, err := gzip.NewReader(bytes.NewReader(message))
+	if err != nil {
+		h.logger.Error("error creating gzip reader", zap.Error(err))
+		return resp, nil, err
+	}
+	defer reader.Close()
```

Comment:
does the defer need to happen before the error?

Reply:
If there is an error, the reader that will be returned is `nil`